### ISSUE TITLE
Detach include template resolution from Engine

### ISSUE THREAD

**Issue by lunabunn** [rossmacarthur/upon#15]
It seems that template resolution being tied to the `Engine` is limiting in several ways:

* The docs recommend using `Engine::compile` if the template source is borrowed, but templates created that way cannot ever be used as an include because there is no way to add an already-compiled `Template` to an `Engine`
* It is not possible to have relative-path includes akin to those found in handlebars, liquid, etc.
* It is not possible to control which templates have access to which other templates

Could `Renderer` (and, by extension, the render methods on `Template`) perhaps store a closure for `get_template` that is supplied by the user, instead of templates being stored in `engine.templates`?

**Comment by lunabunn** [comment]
Considering this would be a breaking change, I am left wondering if you would be open to supporting a usecase like this.

If you would, how would the API look? Should this just be a new function altogether (render_from_with_template_fn, render_with_value_template_fn)? That looks like it would complicate the API a lot, but there may be some middle ground that supports this new use case without affecting the original resolution mechanism.

With your input, depending on what approach you would most prefer, I would be more than happy to draft up a PR. Thank you for your work on this great crate!

**Comment by rossmacarthur** [comment]
I am open to breaking changes. I also thought there would be too many render functions if you want to allow all the combinations. I think it would be good to decide on an approach first before putting in a lot of effort. The one idea I had was for the complex render functions to return a struct that can have these settings modified (similar to the builder 
pattern) and perhaps for the more commonly used render functions to just be shorthand for this. The returned struct could just be basically the current non-public `Renderer` struct with some changes to the API.

If we add this feature we will have the following options:
- Context: simple value / serde value / value fn
- Output: string or writer
- Template Fn: from engine or from custom function
- There is also the max include depth setting which could be set at render time not just in the engine.

### API ideas

Render serde context to writer

```rust
template.render(ctx)?.to_writer(w)?;
```

Render using value function, with template function, to a string

```rust
template.render_with_value_fn(value_fn).with_template_fn(template_fn).to_string()?
```

Render from value to a string
```rust
template.render_from(&value).to_string()?
```

Render from value to a writer
```rust
template.render_from(&value).to_writer(w)?
```

Render with a different max include depth

```rust
template.render(ctx)?.with_max_include_depth(12).to_string()?;
```

Not sure of the naming of all of these methods. Particularly `to_string()` since it's obviously already in the std library.

**Comment by lunabunn** [comment]
I agree that naively adding all the combinations would get very clunky very quickly, and so we should definitely decide on an API first if we were to pursue this further.

I am a fan of the builder-pattern-esque API you suggested! However, one thing is 
```rust
template.render_with_value_fn(value_fn).with_template_fn(template_fn).to_string()?
```
This is confusing to me. `with_value_fn` and `with_template_fn` feel too similar, although one is mandatory and one is just an option.

I think perhaps `render`, `render_from`, and `render_from_fn` would work? You mentioned in #18 that `*_from` might not work if we change the bound to `AsRef<Value>`, but I think it's still fine in that it conveys the idea that you take a value/value fn and render the template using/from it.

We could of course continue to bikeshed nomenclature after we start working on the implementation.

Side note: `String` implements `Write`, so I think it would be nice from a maintenance perspective to have `to_string()` here be implemented in terms of `to_writer()` and remove that distinction from the rendering code itself.

**Comment by rossmacarthur** [comment]
I agree with your point about `with_value_fn` and `with_template_fn`. I like the idea of `render`, `render_from`, and `render_from_fn` as being the main three functions on `Template{,Ref}`


**Possible public API for this**

(I've left out lifetimes for now)

```rust
impl Template(,Ref) {
    pub fn render<S>(&self, ctx: S) -> Result<Renderer>
    where
        S: serde::Serialize,;
    
    pub fn render_from<V>(&self, ctx: V) -> Renderer
    where
        V: AsRef<Value>;
    
    pub fn render_from_fn<F>(&self, value_fn: F) -> Renderer
    where
        F: Fn(&[ValueKey<'_>]) -> std::result::Result<Value, String>;
}

impl Renderer {
    pub fn with_max_include_depth(self, depth: usize) -> Renderer;

    pub fn with_template_fn<F>(self, template_fn: F) -> Renderer
    where
        F: Fn(&str) -> Option<Template>;
        
    pub fn to_string(self) -> Result<String>;

    pub fn to_writer<W>(self, writer: W) -> Result<()>
    where
        W: io::Write;
}
```



**With regards to your side note**

I'm not sure I fully understand your point here. `String` implements `fmt::Write` but the `to_writer` should take an `io::Write`. Most of the render logic is already generic over both of these:

```rust
impl {
    pub fn render(&mut self) -> Result<String> {
        let mut s = String::with_capacity(self.template.source.len());
        let mut f = Formatter::with_string(&mut s);
        self.render_impl(&mut f)?;
        Ok(s)
    }

    pub fn render_to_writer<W>(&mut self, writer: W) -> Result<()>
    where
        W: io::Write,
    {
        let mut w = Writer::new(writer);
        let mut f = Formatter::with_writer(&mut w);
        self.render_impl(&mut f)
            .map_err(|err| w.take_err().map(Error::from).unwrap_or(err))
    }

    fn render_impl(&mut self, f: &mut Formatter<'_>) -> Result<()> {
        // ...
    }
}
```

**Comment by rossmacarthur** [comment]
First version of this implemented in 39a5542981ef6492b23a198823c235dcb4968e4e
